Stop removing elements from the priority queue, speeds up flexible algorithms

easbar opened the pull request:
`PriorityQueue#remove` is an expensive operation, especially for large queues as it has O(N) complexity. We use it for Dijkstra and related algorithms, because there is no efficient update operation: We simply remove the old entry and add the updated entry. An alternative approach is to flag the old entry, add a new one and ignore the flagged ones when polling from the queue. @michaz already did this for isochrones some time ago: #2092

We tried this for the other algorithms already in the past, but without success so far: #489, https://github.com/graphhopper/graphhopper/pull/2107, https://github.com/graphhopper/graphhopper/issues/2408

Trying this again I think this change clearly speeds up the flexible (non-CH) algorithms, especially for the edge-based traversal mode. In our measurements edge-based AStar was more than 30% faster and also the alternative algorithms. For LM the queries are ~25% faster and even the LM preparation is ~15% faster now.

Here are the results:

```
# Bavaria before
routing: sum:27.953s, time/call:0.112s, dummy: 614737
routing_alt: sum:12.457s, time/call:2.491s, dummy: 12939
routing_edge: sum:149.028s, time/call:0.596s, dummy: 615013
routing_edge_alt: sum:165.167s, time/call:33.033s, dummy: 12946

# Bavaria now
routing: sum:26.518s, time/call:0.106s, dummy: 614737
routing_alt: sum:10.845s, time/call:2.169s, dummy: 12939
routing_edge: sum:98.296s, time/call:0.393s, dummy: 615013
routing_edge_alt: sum:130.904s, time/call:26.181s, dummy: 12946

# Germany before
Calculated landmarks for 1 subnetworks, took:311s
routingLM4: sum:14.978s, time/call:0.06s, dummy: 1038316
routingLM4_alt: sum:25.455s, time/call:5.091s, dummy: 17288
routingLM4_edge: sum:78.8s, time/call:0.315s, dummy: 1038820
routingLM4_alt_edge: sum:482.417s, time/call:96.483s, dummy: 17839
routingLM8: sum:14.408s, time/call:0.058s, dummy: 1038316
routingLM8_alt: sum:25.885s, time/call:5.177s, dummy: 17288
routingLM8_edge: sum:61.402s, time/call:0.246s, dummy: 1038820
routingLM8_alt_edge: sum:456.623s, time/call:91.325s, dummy: 17839
routingLM12: sum:15.463s, time/call:0.062s, dummy: 1038316
routingLM12_alt: sum:27.118s, time/call:5.424s, dummy: 17288
routingLM12_edge: sum:64.217s, time/call:0.257s, dummy: 1038820
routingLM12_alt_edge: sum:493.755s, time/call:98.751s, dummy: 17839
routingLM16: sum:16.573s, time/call:0.066s, dummy: 1038316
routingLM16_alt: sum:28.179s, time/call:5.636s, dummy: 17288
routingLM16_edge: sum:66.663s, time/call:0.267s, dummy: 1038820
routingLM16_alt_edge: sum:504.28s, time/call:100.856s, dummy: 17839
routingLM8_block_area: sum:23.476s, time/call:0.094s, dummy: 1049107

# Germany now
Calculated landmarks for 1 subnetworks, took:258s
routingLM4: sum:14.042s, time/call:0.056s, dummy: 1038316
routingLM4_alt: sum:20.834s, time/call:4.167s, dummy: 17288
routingLM4_edge: sum:45.134s, time/call:0.181s, dummy: 1038820
routingLM4_alt_edge: sum:362.848s, time/call:72.57s, dummy: 17839
routingLM8: sum:13.902s, time/call:0.056s, dummy: 1038316
routingLM8_alt: sum:21.381s, time/call:4.276s, dummy: 17288
routingLM8_edge: sum:43.512s, time/call:0.174s, dummy: 1038820
routingLM8_alt_edge: sum:359.046s, time/call:71.809s, dummy: 17839
routingLM12: sum:14.877s, time/call:0.06s, dummy: 1038316
routingLM12_alt: sum:21.979s, time/call:4.396s, dummy: 17288
routingLM12_edge: sum:45.994s, time/call:0.184s, dummy: 1038820
routingLM12_alt_edge: sum:405.193s, time/call:81.039s, dummy: 17839
routingLM16: sum:16.06s, time/call:0.064s, dummy: 1038316
routingLM16_alt: sum:23.153s, time/call:4.631s, dummy: 17288
routingLM16_edge: sum:48.906s, time/call:0.196s, dummy: 1038820
routingLM16_alt_edge: sum:402.926s, time/call:80.585s, dummy: 17839
routingLM8_block_area: sum:22.274s, time/call:0.089s, dummy: 104910
```

boldtrn commented:
I did test this as well and can verify your numbers. This is a serious performance improvement for flexible algos 🎉

easbar commented:
> I did test this as well and can verify your numbers. This is a serious performance improvement for flexible algos

Thanks a lot for trying this out! Note there was a bug in combination with LM and fixed start/target edges (i.e. when using curbsides or map matching), that I just fixed in #2581.

boldtrn commented:
Thanks a lot letting me know about this and for improving this case ❤️
